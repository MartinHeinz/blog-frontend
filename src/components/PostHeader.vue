<template>
    <header>

        <h1 class="posttitle" itemprop="name headline">{{title}}</h1>


        <div class="meta">

            <span class="author" itemprop="author" itemscope="" itemtype="http://schema.org/Person">
                <span itemprop="name">{{author}}</span>
            </span>

            <div class="postdate">
                <time :datetime="published | formatDate" itemprop="datePublished">{{ published | formatDate }}</time>
            </div>

            <div class="article-tag">
                <v-icon>fas fa-tag</v-icon>
                <BaseTag v-for="tag in tags"
                              v-bind:url="tag.url"
                              v-bind:name="tag.name"
                              v-bind:key="tag.key">
                </BaseTag>
            </div>

        </div>
    </header>
</template>

<script>
import BaseTag from '@/components/BaseTag.vue';
import axios from 'axios';
import { API_URL } from '@/common/config';

export default {
    name: 'PostHeader',
    components: { BaseTag },
    props: {
        tags: Array,
        title: String,
        author: String,
        published: String,
    },
    watch: {
        $route: 'updateHeaderData',
    },
    methods: {
        updateHeaderData() {
            return axios
                .get(`${API_URL}posts/${this.$route.params.id}`)
                .then((response) => {
                    this.title = response.data.title;
                    this.author = response.data.author;
                    this.published = response.data.posted_on;
                    this.tags = response.data.tags;
                });
        },
    },
};
</script>

<style>

    .article-tag .tag-link {
        background-image: linear-gradient(transparent, transparent 10px, #d480aa 10px, #d480aa);
        background-position: bottom;
        background-size: 100% 6px;
        background-repeat: repeat-x;
    }


    .tag-link a:hover {
        box-shadow: inset 0 0px 0 white, inset 0 -2px 0 #d480aa;
    }

    .tag-link a {
        color: #c9cacc;
        text-decoration: none;
        background-image: linear-gradient(transparent, transparent 5px, #c9cacc 5px, #c9cacc);
        background-position: bottom;
        background-size: 100% 6px;
        background-repeat: repeat-x;
    }

    article header .posttitle {
        margin-top: 0;
        margin-bottom: 0;
        text-transform: none;
        font-size: 1.5em;
        line-height: 1.25;
    }

    .posttitle {
        text-transform: none;
        font-size: 1.5em;
        line-height: 1.25;
    }

    header .meta {
        margin-top: 0;
        margin-bottom: 1rem;
    }
    header .meta * {
        color: #ccc;
        font-size: 0.85rem;
    }

    header .author {
        text-transform: uppercase;
        letter-spacing: 0.01em;
        font-weight: 700;
        margin: 0 5px;
    }

    header .postdate {
        display: inline;
    }

    @media (min-width: 480px) {
        .article-tag {
            display: inline;
        }
        .article-tag:before {
            content: " | ";
        }
    }

    span:first-child {
        margin-left:0;
    }
</style>
