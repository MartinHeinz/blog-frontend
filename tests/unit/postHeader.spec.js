import { createLocalVue, shallowMount } from '@vue/test-utils';
import PostHeader from '@/components/PostHeader.vue';

import moment from 'moment';
import Vuetify from 'vuetify';
import Vue from 'vue';
import Vuex from 'vuex';

const header = {
    title: 'Second Blog Post',
    author: 'Martin',
    published: '0001-01-01T00:00:00Z',
    tags: [
        {
            id: 1,
            created_at: '0001-01-01T00:00:00Z',
            updated_at: '0001-01-01T00:00:00Z',
            deleted_at: null,
            post_id: 1,
            name: 'Python',
        },
        {
            id: 3,
            created_at: '0001-01-01T00:00:00Z',
            updated_at: '0001-01-01T00:00:00Z',
            deleted_at: null,
            post_id: 1,
            name: 'Crypto',
        }],
};

const $route = {
    hash: '',
    params: { id: '2' },
};

describe('PostHeader.vue', () => {
    Vue.use(Vuex);
    let getters;
    let actions;
    let store;

    beforeEach(() => {
        getters = {
            currentPostHeader: () => header,
        };
        actions = {
            fetchPostById: jest.fn(),
        };

        store = new Vuex.Store({
            getters,
            actions,
        });
    });

    it('Sets header attributes of blog post header', () => {
        const localVue = createLocalVue();
        localVue.use(Vuetify);
        localVue.filter('formatDate', value => moment(String(value)).format('ll')); // TODO import this from @/main.js

        const wrapper = shallowMount(PostHeader, {
            localVue,
            store,
            mocks: {
                $route,
            },
        });

        expect(wrapper.html()).toMatchSnapshot();
    });
});
