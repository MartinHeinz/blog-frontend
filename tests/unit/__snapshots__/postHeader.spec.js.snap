// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostHeader.vue Sets header attributes of blog post header 1`] = `
<header>
  <vue-headful title="Martin Heinz - Second Blog Post"></vue-headful>
  <h1 itemprop="name headline" class="posttitle">Second Blog Post</h1>
  <div class="meta"><span itemprop="author" itemscope="itemscope" itemtype="http://schema.org/Person" class="author"><span itemprop="name">Martin</span></span>
    <div class="postdate"><time datetime="0001-01-01T00:00:00Z" itemprop="datePublished">Jan 1, 0001</time></div>
    <div class="article-tag">
      <v-icon-stub>fas fa-tag</v-icon-stub>
      <basetag-stub name="Python"></basetag-stub>
      <basetag-stub name="Crypto"></basetag-stub>
    </div>
  </div>
</header>
`;
